src/login/tools.js

```diff
@@ -0,0 +1,261 @@
+import _ from 'underscore';
+
+import {staticAttributePatterns, otherAttributePatterns} from './consts';
+import {filters} from './filters';
+
```

Comment:
`getLoginFields` should be the first function here. Then, other function should be sorted below `getLoginFields` in the order they appear in the function. So that one can read the file like a book.

Reply:
You'll need to redefine the `const` functions into actual functions to support this.

Reply:
Yeah, this is a tricky one. I know there should be better order of the functions and will probably aim for nesting them into more levels and splitting them to more separate files based on those layers. I usually try to keep the functions in sequence but since I also like to use the arrow syntax and declare functions as constants, the code then does read like a book but written in arabic since it flows from bottom to top... :-)

I suspect hoisting can solve this without having to reverse the order of everything or having to abandon anonymous arrow functions in constants and force the function declaration but not sure it's a good practice... will look into that more

Reply:
I suggest reading https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29#source-code-structure or the book Clean Code itself. It includes some really good arguments for writing code like they suggest.

I was mostly referring to those two principles from the summary:

6. Place functions in the downward direction.
4. Dependent functions should be close.

Reply:
I do intend to get my hands on the book - thanks for the gist meanwhile. I don't feel like giving up on arrow function syntax completely so I'll need to find some compromise there, but maybe the points you mention will persuade me? :-)

Reply:
> This creates a nice flow down the source code module from high level to low level.

I get the book/newspaper analogy (though arabic speaking population would still disagree :-) but I guess it's determined by how the language itself works as even admitted in the book.

**Vertical Ordering**
_2. This is the exact opposite of languages like Pascal, C, and C++ that enforce functions to be defined, or at least declared, before they are used._

Nevertheless, I reversed the flow from upwards to downwards and it still works, even without turning it into function declarations - not quite sure why.